confmap/converter/templateconverter/config.go

```diff
@@ -0,0 +1,121 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package templateconverter // import "go.opentelemetry.io/collector/confmap/converter/templateconverter"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"text/template"
+
+	"gopkg.in/yaml.v3"
+)
+
+type instanceID struct {
+	tmplType string
+	instName string
+}
+
+func newInstanceID(id string) (*instanceID, error) {
+	parts := strings.SplitN(id, "/", 3)
+	switch len(parts) {
+	case 2: // template/type
+		return &instanceID{
+			tmplType: parts[1],
+		}, nil
+	case 3: // template/type/name
+		return &instanceID{
+			tmplType: parts[1],
+			instName: parts[2],
+		}, nil
+	}
+	return nil, fmt.Errorf("'template' must be followed by type")
+}
+
+func (id *instanceID) withPrefix(prefix string) string {
+	if id.instName == "" {
+		return prefix + "/" + id.tmplType
+	}
+	return prefix + "/" + id.tmplType + "/" + id.instName
+}
+
+type templateConfig struct {
+	Receivers  map[string]any             `yaml:"receivers,omitempty"`
+	Processors map[string]any             `yaml:"processors,omitempty"`
+	Pipelines  map[string]partialPipeline `yaml:"pipelines,omitempty"`
+}
+
+type partialPipeline struct {
+	Receivers  []string `yaml:"receivers,omitempty"`
+	Processors []string `yaml:"processors,omitempty"`
+}
+
+func newTemplateConfig(id *instanceID, tmpl *template.Template, parameters any) (*templateConfig, error) {
+	var rendered bytes.Buffer
+	var err error
+	if err = tmpl.Execute(&rendered, parameters); err != nil {
+		return nil, fmt.Errorf("render: %w", err)
+	}
+
+	cfg := new(templateConfig)
+	if err = yaml.Unmarshal(rendered.Bytes(), cfg); err != nil {
+		return nil, fmt.Errorf("malformed: %w", err)
+	}
+
+	if len(cfg.Receivers) == 0 {
+		return nil, errors.New("must have at least one receiver")
```

Comment:
I fully agree with the other use cases as well. I would describe your example (1) as a "processor template", and (2) as an "exporter template", and this PR as supporting "receiver templates".

I think it's really important to look at templates from two distinct points of view:
1. A novice user who benefits from not knowing the details of the template.
2. An experienced user who captures and shares domain knowledge by creating a template.

For the novice user, the purpose of a template is to be an abstraction. In order to do this effectively, I think a template needs to _feel_ as much as possible like a regular component, meaning it has both a class (i.e. receiver, processor, exporter) and type (e.g. filelog, my_template), and its use in the configuration is almost exactly like any other component. If they can think of it as a receiver, then it's much clearer where to define an instance (in the `receivers` section) and where to use it (in a pipeline's `receivers` section).

For the experienced users who create templates, and for us as maintainers, there are other reasons to constrain each template to a particular class. There are several natural differences which I believe would be very difficult to manage otherwise:
1. Requirements for what is optional and required within the template. (A receiver template must contain a receiver because otherwise using the template in place of a receiver would have no effect.)
2. Rules for how a template _instance_ is defined (in the `receivers` section) and how it must be used in the service (in a pipeline's `receivers` section).
3. Logic for how the template is expanded and how it hooks into the global component graph.

Reply:
That's a really insightful answer, thanks for the details. I really like the idea of ascribing a class to templates, I agree with all your points on how it would simplify both the validation and usage of a template. I'm seeing now that classifying templates is already present in the existing implementation, I've left a few comments/questions around the implementation side of that.